src/hmac.rs

```diff
@@ -245,11 +268,91 @@ impl Key {
     pub fn digest_algorithm(&self) -> &'static digest::Algorithm {
         self.ctx_prototype.inner.algorithm()
     }
+
+    /// Calculates the HMAC of `data` using the key in one step.
+    ///
+    /// Use [`sign_parts`] if the data is not in a contiguous memory region. In
+    /// more involved cases, use [`Context`] to calculate HMACs where the input
+    /// is not available at once.
+    ///
+    /// It is generally not safe to implement HMAC verification by comparing the
+    /// return value of `sign` to a tag. Use [`verify`] for verification instead.
+    ///
+    /// [`sign_parts`]: #method.sign_parts
+    /// [`verify`]: #method.verify
+    /// [`Context`]: ./struct.Context.html
+    pub fn sign(&self, data: &[u8]) -> Tag {
+        let mut ctx = Context::with_key(self);
+        ctx.update(data);
+        ctx.sign()
+    }
+
+    /// Verifies that the HACM of `data` with key equal `tag`.
+    ///
+    /// Use [`verify_parts`] if the data is not in a contiguous memory region.
+    ///
+    /// Conceptually, this will calculates the HMAC of `data` using the signing
+    /// key, and verifies whether the resultant value equals `tag`, in one step
+    /// and without making the resulting tag available. The comparison will be
+    /// done in constant time to prevent timing attacks.
+    ///
+    /// [`verify_parts`]: #method.verify_parts
+    pub fn verify(&self, data: &[u8], tag: &[u8]) -> Result<(), error::Unspecified> {
```

Comment:
It was a concious choice, note that `verify_parts` requires its argument to be generic for the iterator but `verify` does not. Then we can avoid generics altogether for `verify`, potentially decreasing mental load—it is not obvious that `&[u8]` can be passed to the `verify_parts` method from its signature alone. Since `verify_parts` is a more advanced method and `Tag` allows one to get it as a byte slice the intent was to move all syntactical difficulty of Rust to `verify_parts` method as well. I thought this outweighed the benefits of a slightly terser call syntax where the `Tag` is provided as a `Tag` as that would not be the most common use case.

Note that this evaluation might differ if the signature of `verify_parts` were to be changed as proposed by your second alternative, which I had not originally considered. What's your conclusion from both arguments?

Reply:
So after looking at this again, I think your argument for simplicity in `verify` also applies to `verify_parts`. Similarly, there is a monomorphisation cost with using generic params, and that gets worse the more you use.

So now I'm leaning towards taking `tag: &[u8]` everywhere, and forcing the caller to just do `tag.as_ref()` like before. It also makes the function easier to understand (the one generic param is already a lot).

Reply:
If you go with `tag: &[u8]`, feel free to resolve/ignore any of my comments about `tag.as_ref()`.